EIPS/eip-7017.md

```diff
@@ -0,0 +1,281 @@
+
+Some of this problems could be solved, but most of them are in the very architecture of their solution. Because of this, a new solution must be implemented to overcome them.
+
+## Reference Implementation
+
+```solidity
+pragma solidity ^0.8.7;
+contract ERC7017 is IERC7017 {
+    /**
+    * @dev Send a notification to an address from the address executing the function
+    * @param to address to send a notification to
+    * @param data to send: (json format, URI recomended): 
+    * {subject, body, image_url (optional),
+    * transaction_request (optional)}
+    */
+    function walletDM (address to, string memory data) external virtual override {
+        emit DirectMsg(msg.sender, to, data);
+    }
+
+    /**
+    * @dev Send a notification to an address from the smart contract
+    * @param to address to send a notification to
+    * @param data to send: (json format, URI recomended): 
+    * {subject, body, image_url (optional),
+    * transaction_request (optional)}
+    */
+    function contractDM (address to, string memory data) external virtual override {
+        emit DirectMsg(address(this), to, data);
+    }
+    
+    /**
+    * @dev Send a general notification from the address executing the function
+    * @param data to broadcast: (json format, URI recomended):
+    * {subject, body, attention_level (1, 2 or 3), image_url (optional),
+    * transaction_request (optional)}
+    */
+    function walletBroadcast (string memory data) external virtual override {
+        emit BroadcastMsg (msg.sender, data);
+    }
+
+    /**
+    * @dev Send a general notification from the address executing the function
+    * @param data to broadcast: (json format, URI recomended):
+    * {subject, body, attention_level (1, 2 or 3), image_url (optional),
+    * transaction_request (optional)}
+    */
+    function contractBroadcast (string memory data) external virtual override {
+        emit BroadcastMsg (address(this), data);
+    }
+}
+```
+
+## Security Considerations
+
+### Impersonation
+
+There are security considerations for the proposal regarding threats.
+
+The first is related to how the clients will interpret and filter the events. There can be malicious contracts that implement the notification interface but whose functions emit events where the sender is neither the message sender nor the contract. They could try to impersonate another address by changing the sender.
+
+Here is an example of a malitious implementation of a BroadcastMsg event impersonating Vitalik Buterin:
+
+```solidity
+emit BroadcastMsg (0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045, message);
+```
+
+This is why the message sender in the log's topic must allways be verified against the transaction "from" or transaction "to", depending on the notification type (wallet or contract respectively).
```

Comment:
of course, another method would be to have a "known" trusted deployed instance of this interface, which would be hardcoded in the client-side software. that way the client code could avoid performing this check explicitly, since it trusts the contract. this seems like the route that'd be taken in practice for messaging software.